src/test/java/plugins/GitLabPluginTest.java

```diff
@@ -49,307 +58,560 @@ public class GitLabPluginTest extends AbstractJUnitTest {
+        var params = new MergeRequestParams()
+                .withSourceBranch(sourceBranch)
+                .withTargetBranch(targetBranch)
+                .withTitle(mrTitle);
+        return waitFor()
+                .withMessage("Creating merge request '%s'", mrTitle)
+                .withTimeout(Duration.ofSeconds(time.seconds(GITLAB_API_RETRY_TIMEOUT_SECONDS)))
+                .until(() -> {
+                    try {
+                        return gitlabapi.getMergeRequestApi().createMergeRequest(project, params);
+                    } catch (Exception e) {
+                        try {
+                            var filter = new MergeRequestFilter().withProjectId(project.getId());
+                            return gitlabapi.getMergeRequestApi().getMergeRequests(filter).stream()
+                                    .filter(mr -> mrTitle.equals(mr.getTitle()))
+                                    .findFirst()
+                                    .orElse(null);
+                        } catch (Exception ex) {
+                            return null;
+                        }
+                    }
+                });
+    }
+
+    private void awaitBranchAvailabilityViaApi(GitLabApi gitlabapi, Long projectId, String branchName) {
+        long startTime = System.currentTimeMillis();
+        waitFor()
+                .withMessage("Waiting for GitLab branch '%s' to be available", branchName)
+                .withTimeout(Duration.ofSeconds(time.seconds(GITLAB_API_RETRY_TIMEOUT_SECONDS)))
+                .until(() -> {
+                    try {
+                        var branch = gitlabapi.getRepositoryApi().getBranch(projectId, branchName);
+                        return branch != null && branch.getCommit() != null;
+                    } catch (Exception e) {
+                        return false;
+                    }
+                });
+        System.out.println("GitLab branch '" + branchName + "' wait: "
+                + Duration.ofMillis(System.currentTimeMillis() - startTime));
+    }
+
+    private WorkflowJob awaitBranchDiscoveryViaJenkins(WorkflowMultiBranchJob multibranchJob, String branchName) {
+        long startTime = System.currentTimeMillis();
+        WorkflowJob job = waitFor()
+                .withMessage("Waiting for Jenkins to discover branch '%s'", branchName)
+                .withTimeout(Duration.ofSeconds(time.seconds(BRANCH_INDEXING_TIMEOUT_SECONDS)))
+                .pollingEvery(Duration.ofSeconds(POLLING_INTERVAL_SECONDS))
+                .until(() -> {
+                    try {
+                        try {
+                            WorkflowJob existingJob = multibranchJob.getJob(branchName);
+                            existingJob.url("").openStream().close();
+                            return existingJob;
+                        } catch (Exception e) {
+                        }
```

Comment:
Probably another way to write this for more debuggable failures.

Reply:
somehow improved